apache/client/test/lauth/api_client_test.cpp

```diff
@@ -35,7 +35,7 @@ TEST(ApiClient, HttpRequestByApiClientIsCorrect) {
   api_client.authorize(req);
 }
 
-TEST(ApiClient, ResponseOfAllowedReturnsTrue) {
+TEST(ApiClient, DeterminationAllowedReturnsTrue) {
```

Comment:
This name is no longer representative of the test case. We are not concerned with the comparison/boolean. We are only concerned with the correct deserialization/conversion of the JSON response.

These tests should really be moved to a test file for the struct/conversions -- they have nothing to do with the ApiClient. That will also simplify the setup (just a string in, no concept of HTTP, API, or "body").

Then we can pare back the ApiClient test to concern only its behavior... That it calls the HttpClient correctly given its inputs and that it returns an AuthorizationResult. This is really a collaboration test, but since we can't "expect" a call to the conversion function with the response body, we have to do one of two things:

1. Have two tests that have different responses to verify that the body is really being used (effective duplication of the serialization tests)
2. Inject a std::function (or object) that handles conversion and expect that to be called

This is a limitation of C++ (not GMock); [you cannot mock free functions directly](https://google.github.io/googletest/gmock_cook_book.html#mocking-free-functions).

I think the first option is the right one for the moment. We can try to make a template function that does the parse-and-convert, but I think there will be some messing around with syntax to get it clean.

The idea is that the ApiClient would, by default, wrap our parse-and-convert in a std::function. We should only have to write one, and it would use json::parse and basic_json::get as usual, just generically. Then we could inject a mock function and verify that it is called and that our mocked value is returned. That would flatten the tests fully verify ApiClient::authorize to one example (that it delegates everything "interesting"), without overspecifying other code's responsibilities.